src/rabbit_auth_backend_oauth2.erl

```diff
@@ -77,28 +80,50 @@ user_login_authorization(Username, AuthProps) ->
 
 check_vhost_access(#auth_user{impl = DecodedToken},
                    VHost, _AuthzData) ->
+    TokenScopes = get_scopes(DecodedToken),
+    UserScopes = case check_picklock(TokenScopes) of
+        true ->
+            get_picklock_scopes(DecodedToken);
+        false -> 
+            TokenScopes
+    end,
     with_decoded_token(DecodedToken,
         fun() ->
-            Scopes      = get_scopes(DecodedToken),
-            ScopeString = rabbit_oauth2_scope:concat_scopes(Scopes, ","),
+            UserScopes,
+            ScopeString = rabbit_oauth2_scope:concat_scopes(UserScopes, ","),
             rabbit_log:debug("Matching virtual host '~s' against the following scopes: ~s", [VHost, ScopeString]),
-            rabbit_oauth2_scope:vhost_access(VHost, Scopes)
+            rabbit_oauth2_scope:vhost_access(VHost, UserScopes)
         end).
 
 check_resource_access(#auth_user{impl = DecodedToken},
                       Resource, Permission, _AuthzContext) ->
+
+    TokenScopes = get_scopes(DecodedToken),
+    UserScopes = case check_picklock(TokenScopes) of
+        true ->
+            get_picklock_scopes(DecodedToken);
+        false -> 
+            TokenScopes
+        end,
+
     with_decoded_token(DecodedToken,
         fun() ->
-            Scopes = get_scopes(DecodedToken),
-            rabbit_oauth2_scope:resource_access(Resource, Permission, Scopes)
+            rabbit_oauth2_scope:resource_access(Resource, Permission, UserScopes)
         end).
 
 check_topic_access(#auth_user{impl = DecodedToken},
                    Resource, Permission, Context) ->
+    TokenScopes = get_scopes(DecodedToken),
```

Comment:
Not something that is realistic to address in this PR but I cannot help but think about this looking at this new version of this function.

This code will be executed for every message going through the system. We are adding several function calls here and a few conditions. We should consider computing these once at initialization, scopes of a token do not change over its lifetime. We do support token refreshes, though, so cache invalidation is a problem here.

@acogoluegnes

Reply:
Right, we should try to pre-compute things as much as possible (not in this PR though).